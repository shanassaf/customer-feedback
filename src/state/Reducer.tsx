const Reducer = (state: State, action: Action): State => {
	switch (action.type) {
		case 'ADD_FEEDBACK':
			return {
				...state,
				feedback: [action.feedback, ...state.feedback],
			};
		default:
			return state;
	}
};

export default Reducer;
