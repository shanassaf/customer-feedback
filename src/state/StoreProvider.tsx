import React, { createContext, useReducer } from 'react';
import Reducer from './Reducer';

const initialState: State = { comments: [] };

const StoreProvider = ({ children }) => {
	const [state, dispatch] = useReducer(Reducer, initialState);
	return <Context.Provider value={{ state, dispatch }}>{children}</Context.Provider>;
};

export const Context = createContext<ContextType>(null);
export default StoreProvider;
