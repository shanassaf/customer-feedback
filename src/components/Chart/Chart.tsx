import React, { useContext } from 'react';
import styled from 'styled-components';
import { Pie } from 'react-chartjs-2';
import { Context } from '../../state/StoreProvider';

const StyledChart = styled.div``;

export const Chart = (): JSX.Element => {
	const {
		state: { feedback },
	} = useContext(Context);

	const labelsArray = ['0', '1', '2', '3', '4', '5'];
	let dataArray = [0, 0, 0, 0, 0, 0];

	const ratingsArray = feedback.map(({ rating }) => rating);

	ratingsArray.forEach((rating) => {
		dataArray[rating]++;
	});

	const data = {
		labels: labelsArray,
		datasets: [
			{
				label: 'Ratings',
				backgroundColor: ['#FFAACC', '#CCFFCC', '#CCAAFF', '#FFFFCC', '#00FFFF', '#FFBC90'],
				borderColor: 'rgba(0,0,0,1)',
				borderWidth: 2,
				data: dataArray,
			},
		],
	};

	const options = {
		title: {
			display: true,
			text: 'See what others rated us!',
			fontSize: 20,
		},
		legend: {
			display: true,
			position: 'top',
		},
	};

	if (feedback.length === 0) {
		return null;
	}

	return (
		<StyledChart>
			<Pie data={data} options={options} />
		</StyledChart>
	);
};
