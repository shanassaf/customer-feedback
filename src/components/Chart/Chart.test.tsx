import React from 'react';
import { render } from '@testing-library/react';
import { Chart } from './Chart';
import { Context } from '../../state/StoreProvider';

jest.mock('react-chartjs-2', () => ({
	Pie: () => <div>Pie Chart</div>,
}));

describe('Chart', () => {
	const state = {
		feedback: [
			{
				name: 'name',
				email: 'email',
				comment: 'text',
				rating: '1',
				uuid: 'uuid',
			},
		],
	} as State;
	test('component renders correctly', () => {
		const { asFragment } = render(
			<Context.Provider value={{ dispatch: () => {}, state }}>
				<Chart />
			</Context.Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});
	test('component does not render if no feedback has been provided', () => {
		const { asFragment } = render(
			<Context.Provider value={{ dispatch: () => {}, state: { feedback: [] } }}>
				<Chart />
			</Context.Provider>
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
