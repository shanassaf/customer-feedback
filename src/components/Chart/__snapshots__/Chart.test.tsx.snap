// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chart component does not render if no feedback has been provided 1`] = `<DocumentFragment />`;

exports[`Chart component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-bdnylx hRRney"
  >
    <div>
      Pie Chart
    </div>
  </div>
</DocumentFragment>
`;
